File: designui/src/core/orset.ts
import { generate } from "short-uuid";
export class OrSet{
    added:Set<string>
    removed:Set<string>

    constructor(){
        this.added = new Set();
        this.removed = new Set();
    }

    lookup(element:string):boolean {
        for(const elem of this.added){
            if(elem.split("||")[0] == element){
                if(!this.removed.has(elem)) return true;
            }
        }
        return false;
    }

    snapshot():Set<string>{

        const result = new Set<string>()
        for(const item of this.added){
            if(!this.removed.has(item)){
                result.add(item.split("||")[0])
            }
        }
        return result
    }

    add(element:string){
        this.added.add(element + "||" + generate())
    }

    remove(element:string):void{
        if(!this.lookup(element)){
            return
        }

        const subtractionSet = new Set<string>()
        for(const elem of this.added){
            if(elem.split("||")[0] === element){
                subtractionSet.add(elem)
            }
        }


        for(const elem of subtractionSet){
            this.added.delete(elem)
            this.removed.add(elem)
        }
    }


    merge(other:OrSet){
        for(const item of other.added){
            this.added.add(item)
        }

        for(const item of other.removed){
            this.removed.add(item)
        }
    }

    deepcopy():OrSet{
        let other = new OrSet()
        
        for(const item of this.added){
            other.added.add(item)
        }

        for(const item of this.removed){
            other.removed.add(item)
        }

        return other
    }

    stateSnapshot(): { added: string[]; removed: string[] } {
        return {
            added: Array.from(this.added),
            removed: Array.from(this.removed)
        };
    }

    mergeStateSnapshot(other: { added: string[]; removed: string[] }): void {
        for (const elem of other.added) {
            this.added.add(elem);
        }

        for (const elem of other.removed) {
            this.removed.add(elem);
        }
    }
}

File: designserver/core/orset.py
import shortuuid

class OrSet:
    def __init__(self):
        self.added = set()
        self.removed = set()
    
    def lookup(self, element:str):
        for elem in self.added.difference(self.removed):
            if elem.split("||")[0] == element:
                return True
        return False
    
    def snapshot(self):
        result = set()
        for elem in self.added.difference(self.removed):
            result.add(elem.split("||")[0])
        return result
    
    def add(self, element):
        self.added.add(element + "||" + shortuuid.uuid())
    
    def remove(self, element):
        if not self.lookup(element):
            return False
        
        subtraction_set = set()
        for elem in self.added:
            if elem.split("||")[0] == element:
                subtraction_set.add(elem)
        
        self.added.difference_update(subtraction_set)
        self.removed.update(subtraction_set)
    
    def merge(self, other:OrSet):
        self.added.update(other.added)
        self.removed.update(other.removed)
    

    def state_snapshot(self):
        return {
            "added": list(self.added),
            "removed": list(self.removed)
        }
    
    def merge_state_snapshot(self, other):
        for elem in other['added']:
            self.added.add(elem)
        for elem in other['removed']:
            self.removed.add(elem)
